internal/jobcontainers/logon.go

```diff
@@ -1,35 +1,133 @@
+	"unsafe"
 
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/log"
 	"github.com/Microsoft/hcsshim/internal/winapi"
 	"github.com/pkg/errors"
 	"golang.org/x/sys/windows"
 )
 
-// processToken returns a user token for the user specified by `user`. This should be in the form
-// of either a DOMAIN\username or just username.
-func processToken(user string) (windows.Token, error) {
+func randomPswd() (*uint16, error) {
+	g, err := guid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+	return windows.UTF16PtrFromString(g.String())
+}
+
+func groupExists(groupName string) bool {
+	var p *byte
+	if err := winapi.NetLocalGroupGetInfo(
+		"",
+		groupName,
+		1,
+		&p,
+	); err != nil {
+		return false
+	}
+	defer windows.NetApiBufferFree(p)
+	return true
+}
+
+// makeLocalAccount creates a local account with the passed in username and a randomly generated password.
+// If `groupName` is not an empty string the user will be added to group if it exists.
+func makeLocalAccount(ctx context.Context, user, groupName string) (*uint16, error) {
+	// Create a local account with a random password
+	pswd, err := randomPswd()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate random password: %w", err)
+	}
+
+	userUTF16, err := windows.UTF16PtrFromString(user)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode username to UTF16: %w", err)
+	}
+
+	usr1 := &winapi.UserInfo1{
+		Name:     userUTF16,
+		Password: pswd,
+		Priv:     winapi.USER_PRIV_USER,
+		Flags:    winapi.UF_NORMAL_ACCOUNT | winapi.UF_DONT_EXPIRE_PASSWD,
+	}
+	if err := winapi.NetUserAdd(
+		"",
+		1,
+		(*byte)(unsafe.Pointer(usr1)),
+		nil,
+	); err != nil {
+		return nil, fmt.Errorf("failed to create user %s: %w", user, err)
+	}
+
+	log.G(ctx).WithField("username", user).Debug("Created local user account for job container")
+
+	sid, _, _, err := windows.LookupSID("", user)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup SID for user %q: %w", user, err)
+	}
+
+	sids := []winapi.LocalGroupMembersInfo0{{Sid: sid}}
+	if err := winapi.NetLocalGroupAddMembers(
+		"",
+		groupName,
+		0,
+		(*byte)(unsafe.Pointer(&sids[0])),
+		1,
+	); err != nil {
+		return nil, fmt.Errorf("failed to add user %q to the %q group: %w", user, groupName, err)
+	}
+
+	return pswd, nil
+}
+
+// processToken verifies first whether userOrGroup is a username or group name. If it's a valid group name,
+// a temporary local user account will be created and added to the group and then the token for the user will
+// be returned. If it is not a group name then the user will logged into and the token will be returned.
+func (c *JobContainer) processToken(ctx context.Context, userOrGroup string) (windows.Token, error) {
 	var (
 		domain   string
 		userName string
 		token    windows.Token
 	)
 
-	split := strings.Split(user, "\\")
+	if userOrGroup == "" {
+		return 0, errors.New("empty username or group name passed")
+	}
+
+	if groupExists(userOrGroup) {
```

Comment:
could there be a username on the system that is the same as a groupname on the system?

Reply:
No afaik in Windows they share the same "namespace" if you will. 
```
net.exe localgroup hpc-users-test /add
net.exe user hpc-users-test somepassword /add
```
after one another fail. If you change the username for the second it functions